### PR TITLE
CI: switch from macOS-12 runners to macOS-13

### PR DESCRIPTION
**Pull request by tronical**
macOS 12 will be phased out later this year and start failing randomly:

https://github.com/actions/runner-images/issues/10721

macOS-13 runners are still on x86-64, so this should be a no-op.

### CODE REVIEW

File: internal/compiler/generator/cpp.rs [hunk 1]
```diff
@@ -3120,22 +3120,30 @@ fn compile_expression(expr: &llr::Expression, ctx: &EvaluationContext) -> String
...
-                    '/' => "/(float)",
-                    _ => op.encode_utf8(&mut buffer),
-                },
-            )
+            let lhs_str = compile_expression(lhs, ctx);
+            let rhs_str = compile_expression(rhs, ctx);
+
+            let lhs_ty = lhs.ty(ctx);
+
+            if (lhs_ty == Type::Float32 || lhs_ty.as_unit_product().is_some()) && (*op == '=' || *op == '!') {
```
Comment (ogoffart) [inline internal/compiler/generator/cpp.rs:3128]:
```suggestion
            if lhs_ty.as_unit_product().is_some() && (*op == '=' || *op == '!') {
```

Float32 is a unit product (an empty one, but an unit product nevertheless)

Comment (tronical) [reply internal/compiler/generator/cpp.rs:3128]: Great :)

File: internal/compiler/generator/cpp.rs [hunk 2]
```diff
@@ -3120,22 +3120,30 @@ fn compile_expression(expr: &llr::Expression, ctx: &EvaluationContext) -> String
...
-                },
-            )
+            let lhs_str = compile_expression(lhs, ctx);
+            let rhs_str = compile_expression(rhs, ctx);
+
+            let lhs_ty = lhs.ty(ctx);
+
+            if lhs_ty.as_unit_product().is_some() && (*op == '=' || *op == '!') {
+                let op = if *op == '=' { "<" } else { ">=" };
+                format!("(std::abs(float({lhs_str} - {rhs_str})) {op} std::numeric_limits<float>::epsilon())")
```
Comment (ogoffart) [inline internal/compiler/generator/cpp.rs:3130]: Note that the rust and interpreter code still do `==`  So this introduce a small difference between languages.

Comment (tronical) [reply internal/compiler/generator/cpp.rs:3130]: Urgh, indeed. That's not good. I'll try a change for that as well then.

**Review by ogoffart — COMMENTED** [review]
Also https://github.com/slint-ui/slint/blob/4718d5b0502d74c6e7ff0a746a89f44cb2fa7fdd/internal/interpreter/eval.rs#L252-L253

does this patch mean that macos-12 is not longer supported?

**Comment by tronical** [comment]
> Also
> 
> https://github.com/slint-ui/slint/blob/4718d5b0502d74c6e7ff0a746a89f44cb2fa7fdd/internal/interpreter/eval.rs#L252-L253

This is the bit I mentioned with `ParitialEq` in the commit message: It's already using `approx_eq`.

> does this patch mean that macos-12 is not longer supported?

If GitHub removes that runner by 1.9 - which seems likely, then yes.

**Review by ogoffart — APPROVED** [review]

**Pull request merged**